File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {

    let left = 0, right = nums.length -1;
    while(left < right){
        
        let mid = Math.floor((left+right)/2);
        if(nums[mid] == target){
            return mid;
        } else if(nums[mid] > nums[right]){
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    let lowestIdx = left;
    return searchOff(nums, 0, nums.length-1,target, lowestIdx);
}

const searchOff = (nums, low, high, target, offset) =>{
    if(low > high) return -1;
    
    let mid = Math.floor((low + high)/2);
    let off = (mid + offset) % nums.length;
    
    if(nums[off] == target)  return off;
    
    if(target < nums[off]){
        return searchOff(nums, low, mid - 1, target, offset);
    } else {
        return searchOff(nums, mid + 1, high, target, offset);
    }
}

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        int l = 0;
        int r = nums.length-1;
        
        while(l <= r){
            int mid = l + (r-l)/2;
            int left = nums[l];
            int middle = nums[mid];
            int right = nums[r];
            
            if(middle == target){
                return mid;
            }
            
            if(left <= middle){
                if(left <= target && target <= middle){
                    r = mid - 1;
                } else {
                    l = mid + 1;
                }
            } else {
                if(middle <= target && target <= right){
                    l = mid + 1;
                } else {
                    r = mid - 1;
                }
            }
        }
        return -1;
    }
}